packages/server/modules/gatekeeper/services/subscriptions/upgradeWorkspaceSubscription.ts

```diff
@@ -0,0 +1,381 @@
+
+export const upgradeWorkspaceSubscriptionFactoryOld =
+  ({
+    getWorkspacePlan,
+    getWorkspacePlanProductId,
+    getWorkspacePlanPriceId,
+    getWorkspaceSubscription,
+    reconcileSubscriptionData,
+    updateWorkspaceSubscription,
+    countWorkspaceRole,
+    upsertWorkspacePlan
+  }: {
+    getWorkspacePlan: GetWorkspacePlan
+    getWorkspacePlanProductId: GetWorkspacePlanProductId
+    getWorkspacePlanPriceId: GetWorkspacePlanPriceId
+    getWorkspaceSubscription: GetWorkspaceSubscription
+    reconcileSubscriptionData: ReconcileSubscriptionData
+    updateWorkspaceSubscription: UpsertWorkspaceSubscription
+    countWorkspaceRole: CountWorkspaceRoleWithOptionalProjectRole
+    upsertWorkspacePlan: UpsertPaidWorkspacePlan
+  }) =>
+  async ({
+    workspaceId,
+    targetPlan,
+    billingInterval
+  }: {
+    workspaceId: string
+    targetPlan: PaidWorkspacePlans
+    billingInterval: WorkspacePlanBillingIntervals
+  }) => {
+    const workspacePlan = await getWorkspacePlan({ workspaceId })
+
+    if (!workspacePlan) throw new WorkspacePlanNotFoundError()
+    switch (workspacePlan.name) {
+      case 'unlimited':
+      case 'academia':
+      case 'starterInvoiced':
+      case 'plusInvoiced':
+      case 'businessInvoiced':
+      case 'free': // TODO: Don't we want to allow upgrades from free to paid?
+        throw new WorkspaceNotPaidPlanError()
+      case 'starter':
+      case 'plus':
+      case 'business':
+      case 'team':
+      case 'pro':
+        break
+      default:
+        throwUncoveredError(workspacePlan)
+    }
+
+    switch (workspacePlan.status) {
+      case 'canceled':
+      case 'cancelationScheduled':
+      case 'paymentFailed':
+      case 'trial':
+      case 'expired':
+        throw new WorkspaceNotPaidPlanError()
+      case 'valid':
+        break
+      default:
+        throwUncoveredError(workspacePlan)
+    }
+
+    const workspaceSubscription = await getWorkspaceSubscription({ workspaceId })
+    if (!workspaceSubscription) throw new WorkspaceSubscriptionNotFoundError()
+
+    const planOrder: Record<PaidWorkspacePlans, number> = {
+      // old
+      business: 3,
+      plus: 2,
+      starter: 1,
+      // new
+      team: 1,
+      pro: 2
+    }
+
+    if (isNewPlanType(workspacePlan.name) || isNewPlanType(targetPlan)) {
+      throw new NotImplementedError()
+    }
+
+    const planCheckers = [isNewPlanType, isOldPaidPlanType]
+    for (const isSpecificPlanType of planCheckers) {
+      const oldPlanFitsSchema = isSpecificPlanType(workspacePlan.name)
+      const newPlanFitsSchema = isSpecificPlanType(targetPlan)
+      if (xor(oldPlanFitsSchema, newPlanFitsSchema)) {
+        throw new WorkspacePlanUpgradeError(
+          'Attempting to switch between incompatible plan types'
+        )
+      }
+    }
+
+    if (isNewPlanType(targetPlan) || isNewPlanType(workspacePlan.name)) {
+      // Needs custom logic below for seats
+      throw new NotImplementedError()
```

Comment:
all of these NotImplemented throws that were added to show where new plans need logic changes should be removed now that we have the actual new implementation, no?

Reply:
I wanted to leave the old implementation exactly as it was to avoid missing some changes in the CR.